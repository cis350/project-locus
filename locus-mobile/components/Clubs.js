import React, { useState, useEffect } from 'react';
import {
  View, Text, StyleSheet, ScrollView, TouchableOpacity, Modal, TouchableHighlight, TextInput, Alert,
} from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import { useIsFocused } from '@react-navigation/native';
import {
  createClub, getUserId, getUserClubs, getSpecificClub, joinClub,
} from '../modules/api';

export default function Clubs({ route, navigation }) {
  const [userClubs, setUserClubs] = useState([]);
  const [modalVisible, setModalVisible] = useState(false);
  const [joinClubModalVisible, setJoinClubModalVisible] = useState(false);
  const [newClubName, setNewClubName] = useState('');
  const [newClubPassword, setNewClubPassword] = useState('');
  const [joinClubName, setJoinClubName] = useState('');
  const [joinClubPassword, setJoinClubPassword] = useState('');
  const { user } = route.params;

  const isFocused = useIsFocused();
  useEffect(() => {
    async function initialize() {
      setUserClubs((await getUserClubs(user.email)).jsonContent);
    }
    initialize();
  }, [isFocused]);

  // set up the view for all the clubs that the user is in
  const displayClubs = [];
  for (let i = 0; i < userClubs.length; i += 1) {
    displayClubs.push(
      <TouchableOpacity style={styles.club} key={`userClub${i}`} onPress={() => showClub(userClubs[i].clubName)}>
        <Text style={styles.clubText}>{userClubs[i].clubName}</Text>
        <Text style={styles.clubText}>Role: {userClubs[i].role}</Text>
        <Ionicons
          style={{ color: 'white', textAlign: 'center', paddingVertical: 20 }}
          name="settings"
          size={24}
        />
      </TouchableOpacity>,
    );
  }

  async function showClub(clubName) {
    const club = (await getSpecificClub(clubName)).jsonContent;
    navigation.navigate('Club', { club });
  }

  async function handleCreateClub() {
    const masterId = await getUserId(user.email);
    const response = await createClub(newClubName, masterId, newClubPassword);
    if (response.status === 200) Alert.alert('Club Created!');
    if (response.status !== 200) Alert.alert('Creation Failed');
    setUserClubs((await getUserClubs(user.email)).jsonContent);
    setNewClubName('');
    setNewClubPassword('');
    setModalVisible(false);
  }

  async function handleJoinClub() {
    const response = await joinClub(joinClubName, user.email, joinClubPassword);
    setUserClubs((await getUserClubs(user.email)).jsonContent);
    setJoinClubName('');
    setJoinClubPassword('');
    if (response.status === 201) Alert.alert('Club Joined!');
    if (response.status !== 201) Alert.alert('Join Club Failed');
    setJoinClubModalVisible(false);
  }

  return (
    <ScrollView>
      <View style={styles.container}>
        {/* modal for creating clubs */}
        <View style={styles.centeredView}>
          <Modal
            animationType="slide"
            transparent={false}
            visible={modalVisible}
          >
            <View style={styles.centeredView}>
              <Text style={{ fontSize: 40, marginBottom: 30 }}>Create Your Club</Text>
              <TextInput
                style={styles.input}
                placeholder="Club Name"
                placeholderTextColor="grey"
                onChangeText={setNewClubName}
                value={newClubName}
              />
              <TextInput
                style={styles.input}
                placeholder="Password"
                placeholderTextColor="grey"
                onChangeText={setNewClubPassword}
                value={newClubPassword}
                // eslint-disable-next-line react/jsx-boolean-value
                secureTextEntry={true}
              />
              <TouchableHighlight style={styles.button} underlayColor="#33E86F" onPress={() => handleCreateClub()}>
                <Text style={{ fontSize: 20 }}>Create</Text>
              </TouchableHighlight>
              <TouchableHighlight style={styles.cancelButton} underlayColor="#33E86F" onPress={() => setModalVisible(false)}>
                <Text style={{ fontSize: 20 }}>Cancel</Text>
              </TouchableHighlight>
            </View>
          </Modal>
        </View>
        <View style={styles.centeredView}>
          <Modal
            animationType="slide"
            transparent={false}
            visible={joinClubModalVisible}
          >
            <View style={styles.centeredView}>
              <Text style={{ fontSize: 40, marginBottom: 30 }}>Join Club</Text>
              <TextInput
                style={styles.input}
                placeholder="Club Name"
                placeholderTextColor="grey"
                onChangeText={setJoinClubName}
                value={joinClubName}
              />
              <TextInput
                style={styles.input}
                placeholder="Password"
                placeholderTextColor="grey"
                onChangeText={setJoinClubPassword}
                value={joinClubPassword}
                // eslint-disable-next-line react/jsx-boolean-value
                secureTextEntry={true}
              />
              <TouchableHighlight style={styles.button} underlayColor="#33E86F" onPress={() => handleJoinClub()}>
                <Text style={{ fontSize: 20 }}>Join</Text>
              </TouchableHighlight>
              <TouchableHighlight style={styles.cancelButton} underlayColor="#33E86F" onPress={() => setJoinClubModalVisible(false)}>
                <Text style={{ fontSize: 20 }}>Cancel</Text>
              </TouchableHighlight>
            </View>
          </Modal>
        </View>
        <Text style={{ fontSize: 24 }}>Which Club Needs Work?</Text>
        <TouchableHighlight style={styles.button} underlayColor="#33E86F" onPress={() => setModalVisible(true)}>
          <Text style={{ fontSize: 20 }}>Create Club</Text>
        </TouchableHighlight>
        <TouchableHighlight style={styles.button} underlayColor="#33E86F" onPress={() => setJoinClubModalVisible(true)}>
          <Text style={{ fontSize: 20 }}>Join Club</Text>
        </TouchableHighlight>
        <View style={styles.clubContainer}>
          {displayClubs}
        </View>
      </View>
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    alignItems: 'center',
  },
  centeredView: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: 20,
  },
  clubContainer: {
    flexDirection: 'column',
    justifyContent: 'space-between',
    backgroundColor: '#B5E48C',
    paddingVertical: 30,
    paddingHorizontal: 20,
    borderRadius: 10,
    marginVertical: 30,
  },
  club: {
    flexDirection: 'column',
    justifyContent: 'space-between',
    alignItems: 'center',
    backgroundColor: '#6A9B72',
    width: 300,
    height: 200,
    marginVertical: 10,
    borderRadius: 10,
  },
  clubText: {
    textAlign: 'center',
    fontSize: 24,
    color: 'white',
    marginVertical: 5,
  },
  input: {
    backgroundColor: 'white',
    borderColor: 'gray',
    width: 250,
    borderWidth: 1,
    borderRadius: 10,
    padding: 10,
    marginVertical: 5,
  },
  button: {
    backgroundColor: '#6A9B72',
    borderRadius: 10,
    paddingVertical: 10,
    width: 150,
    justifyContent: 'center',
    alignItems: 'center',
    margin: 10,
    paddingHorizontal: 5,
  },
  cancelButton: {
    backgroundColor: 'red',
    borderRadius: 10,
    paddingVertical: 10,
    width: 150,
    justifyContent: 'center',
    alignItems: 'center',
    margin: 10,
    paddingHorizontal: 5,
  },
});
