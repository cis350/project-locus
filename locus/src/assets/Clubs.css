.club-header {
    font-family: Arial;
    margin-left: 50px;
}

.club-table {
    background-color: #B5E48C;
    border-radius: 10px;
    text-align:center;
    margin-left: 50px;
    margin-right: 50px;
}

.club-item {
    margin: auto;
}

.club-button {
    font-family: Arial;
    background-color: #6A9B72;
    font-Size: 24px;
    width: 90%;
    border-Color: #B1DFAD;
    margin-left: 20px;
    margin-right: 20px;
    margin-top: 10px;
    margin-bottom: 10px;
}

.club-button-info {
    width: 100%;
    display: inline-flex;
    text-align: center;
    list-style-type: none;
    margin: 0;
    padding: 0px 10px;
}

.club-button-info-item {
    display:inline;
    padding: 10px 20px;
}

.add-club-form {
    text-align: center;
}

.club-form {
    margin: auto;
}

.add-club-button {
    width: 20%;
    margin: auto;
}