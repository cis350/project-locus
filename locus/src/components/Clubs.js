import React, { useState, useRef } from 'react';
import {
  Button, Form, Container, Row, Col, Stack,
} from 'react-bootstrap';
import '../assets/Clubs.css';
import {
  getUserFullName, getUserClubs, joinClub, getClub,
} from '../modules/storage';
import { v4 as uuidv4 } from 'uuid';

function Clubs({ userEmail }) {
  // initial clubs user is in
  const initState = [];
  const userClubs = getUserClubs(userEmail);
  for (let i = 0; i < userClubs.length; i++) {
    const club = getClub(userClubs[i]);
    const newClub = {
      clubItemName: userClubs[i],
      masterItemName: club.master,
    };
    initState.push(newClub);
  }
  const [clubsArray, addClub] = useState(initState);

  const clubName = useRef('');
  const masterName = useRef('');

  const handleClubs = () => {
    const clubValues = joinClub(userEmail, clubName.current, masterName.current, uuidv4())
    if (clubValues){
      const newClub = {
        clubItemName: clubName.current,
        masterItemName: masterName.current,
      };
      addClub([...clubsArray, newClub]);
    } else {
      alert('Club exists, wrong master');
    }
  };

  const makeClubName = (e) => {
    clubName.current = e.target.value;
  };

  const makeMasterName = (e) => {
    masterName.current = e.target.value;
  };

  return (
    <div>
      <Container fluid>
        <Stack gap={20}>
          <Row>
            <h1 className="club-header">
              Which Club Needs Work, &nbsp;
              {getUserFullName(userEmail)}
              ?
            </h1>
          </Row>
          <div className="club-table">
            {clubsArray.map((clubItem) => (
              <div className="club-item" key={clubItem.clubItemName}>
                <Button className="club-button">
                  <Row>
                    <Col className="d-flex justify-content-center">
                      {clubItem.clubItemName}
                    </Col>
                    <Col className="d-flex justify-content-center">
                      Master: &nbsp;
                      {clubItem.masterItemName}
                    </Col>
                    <Col className="d-flex justify-content-center">
                      Settings
                    </Col>
                  </Row>
                </Button>
              </div>
            ))}
          </div>
          <div className="add-club-form">
            <Row>
              <Form className="club-form">
                <input type="text" onChange={makeClubName} placeholder="Club name" />
              </Form>
              <Form className="club-form">
                <input type="text" onChange={makeMasterName} placeholder="Clubmaster name" />
              </Form>
              <Button className="add-club-button" onClick={handleClubs}>
                Add Club
              </Button>
            </Row>
          </div>
        </Stack>
      </Container>
    </div>
  );
}

export default Clubs;
