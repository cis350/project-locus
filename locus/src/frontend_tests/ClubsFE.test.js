/**
* @jest-environment jsdom
*/
import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import Clubs from '../components/Clubs';

const axios = require('axios');
const MockAdapter = require('axios-mock-adapter');
const domain = 'http://localhost:3306';

// mock useNavigate: https://github.com/remix-run/react-router/issues/7811
const mockedNavigator = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedNavigator,
}));

let mock;
beforeAll(() => {
  mock = new MockAdapter(axios);
});
afterEach(() => {
  mock.reset();
});

test('club snapshot test', () => {
  const component = renderer.create(<Clubs />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
