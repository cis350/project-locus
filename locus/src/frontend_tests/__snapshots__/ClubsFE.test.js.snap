// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clubs snapshot test 1`] = `
<div
  className="d-flex justify-content-center"
>
  Loading...
</div>
`;
